shared/lib/archive-history.js

```diff
@@ -28,22 +28,22 @@ class ArchiveHistory extends History {
     // downstream call of `urlRouter.attach()` which we do when bootstraping
     // everything.
     if (window.history) {
-      let replacingUrl = url;
+      let replacingUrl = window.location.search + url;
       // This is a way to make sure the hash gets cleared out
       if (url === '') {
-        replacingUrl = window.location.pathname;
+        replacingUrl = window.location.pathname + window.location.search;
       }
-      super.replaceUrlSilently(replacingUrl + window.location.search);
+      super.replaceUrlSilently(replacingUrl);
     }
   }
 
   pushUrlSilently(url) {
-    let replacingUrl = url;
+    let replacingUrl = window.location.search + url;
     // This is a way to make sure the hash gets cleared out
     if (url === '') {
-      replacingUrl = window.location.pathname;
+      replacingUrl = window.location.pathname + window.location.search;
     }
-    super.pushUrlSilently(replacingUrl + window.location.search);
+    super.pushUrlSilently(replacingUrl);
```

Comment:
These changes should be split out. This fixes the problem of the hash and query parameter ordering and repeating.

Reply:
Split out to https://github.com/matrix-org/matrix-public-archive/pull/99